### DISCUSSION TITLE
Trouble Installing

### DISCUSSION THREAD

**Discussion by ImShoes** [enhuiz/vall-e#60] [Q&A]
I'm fairly new to python, only been doing it for around a year and a half now and I've never really experienced any problems with installing things through pip until now. After installing PyTorch through pip I tried installing vall-e using the command in the read.me but every time I run it, this error gets thrown.
I'm struggling pretty hard here and I don't know what the problem is.
Anyone possibly know what this could be caused from and how to fix it?
I think its a problem with deepspeed because it seems that the error gets thrown while installing that and trying to manually install deepspeed it doesn't work but I still can't find a way to fix it.

```
Collecting git+https://github.com/enhuiz/vall-e
  Cloning https://github.com/enhuiz/vall-e to c:\users\ian rothenbuhler\appdata\local\temp\pip-req-build-sz_b49ur
  Running command git clone --filter=blob:none --quiet https://github.com/enhuiz/vall-e 'C:\Users\Ian Rothenbuhler\AppData\Local\Temp\pip-req-build-sz_b49ur'
  Resolved https://github.com/enhuiz/vall-e to commit 3476d393d2133fa9b50d5ad999ca13b95fc22060
  Running command git submodule update --init --recursive -q
  Preparing metadata (setup.py) ... done
Requirement already satisfied: coloredlogs>=15.0.1 in c:\users\ian rothenbuhler\appdata\local\programs\python\python310\lib\site-packages (from vall-e==0.0.1.dev20230119082310) (15.0.1)
Collecting deepspeed>=0.7.7
  Using cached deepspeed-0.8.1.tar.gz (759 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [16 lines of output]
      test.c
      LINK : fatal error LNK1181: cannot open input file 'aio.lib'
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\Ian Rothenbuhler\AppData\Local\Temp\pip-install-ar6n3wns\deepspeed_da91508045a346c3adad758cf68af24f\setup.py", line 156, in <module>
          abort(f"Unable to pre-compile {op_name}")
        File "C:\Users\Ian Rothenbuhler\AppData\Local\Temp\pip-install-ar6n3wns\deepspeed_da91508045a346c3adad758cf68af24f\setup.py", line 48, in abort
          assert False, msg
      AssertionError: Unable to pre-compile async_io
      [WARNING] Torch did not find cuda available, if cross-compiling or running with cpu only you can ignore this message. Adding compute capability for Pascal, Volta, and Turing (compute capabilities 6.0, 6.1, 6.2)
      DS_BUILD_OPS=1
       [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
       [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
       [WARNING]  One can disable async_io with DS_BUILD_AIO=0
       [ERROR]  Unable to pre-compile async_io
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by fm-frga** [comment]
The error you're encountering is related to the deepspeed package, specifically with its async_io component, which requires the libaio development library and headers to be installed on your system. The package failed to compile since it could not find the required 'aio.lib' file.

To resolve this issue, you can try the following:

Install the libaio development library on your system. If you're using a Linux-based system, you can use the following commands:

For Ubuntu/Debian:

```
sudo apt-get update
sudo apt-get install libaio-dev
```

For CentOS/RHEL:

```
sudo yum install libaio-devel
```

On Windows, you can try to find a compatible libaio package, or consider using Windows Subsystem for Linux (WSL) and then install the required libraries.

If you've already installed libaio and the problem persists, set the CFLAGS and LDFLAGS environment variables to point to the correct paths where libaio is installed.

If you don't need the async_io functionality, you can try disabling it by setting the DS_BUILD_AIO environment variable to 0:

```
set DS_BUILD_AIO=0
```

Then, retry installing the vall-e package. Note that by doing this, your code will run extremely slower, but it will run.

Keep in mind that the issue is not with pip, but rather with the package you're trying to install. If the issue persists, you may want to raise an issue on the package's GitHub repository or check for any existing issues with similar errors.
